mapie/regression/utils/ccp_phi_function.py

```diff
@@ -0,0 +1,974 @@
+    """
+
+    fit_attributes: Optional[List[str]] = []
+    output_attributes = ["n_in", "n_out", "init_value"]
+
+    def __init__(
+        self,
+        functions: Optional[Union[Callable, Iterable[Callable]]] = None,
+        marginal_guarantee: bool = True,
+        normalized: bool = False,
+    ) -> None:
+        self.functions = functions
+        self.marginal_guarantee = marginal_guarantee
+        self.normalized = normalized
+
+    def _check_transform_parameters(self) -> None:
+        """
+        Check that ``functions_`` are functions that take as input
+        allowed arguments
+        """
+        self.functions_ = self._check_functions()
+
+    def _check_functions(
+            self,
+    ) -> NDArray:
+        """
+        Validate functions for required and optional arguments.
+
+        Raises
+        ------
+        ValueError
+            If no functions are provided and `marginal_guarantee` is False.
+            If functions contain unknown required arguments.
+
+        Warns
+        -----
+        UserWarning
+            If functions contain unknown optional arguments.
+
+        Notes
+        -----
+        This method ensures that the provided functions only use recognized
+        arguments ('X', 'y_pred', 'z'). Unknown optional arguments are allowed,
+        but will always use their default values.
+        """
+        if self.functions is None:
+            self.functions = cast(NDArray, [])
+        elif isinstance(self.functions, Iterable):
+            self.functions = cast(NDArray, self.functions)
+        else:
+            self.functions = cast(NDArray, [self.functions])
+
+        if (len(self.functions) == 0) and not self.marginal_guarantee:
+            raise ValueError("You need to define the `functions` argument "
+                             "with a function or a list of functions, "
+                             "or keep marginal_guarantee argument to True.")
+
+        warn_ind: Dict[str, List[int]] = {}
+        error_ind: Dict[str, List[int]] = {}
+        for i, funct in enumerate(self.functions):
+            assert callable(funct)
+            params = inspect.signature(funct).parameters
+
+            for param, arg in params.items():
+                if (
+                    param not in ["X", "y_pred", "z"]
+                    and param != "disable_marginal_guarantee"
+                ):
+                    if arg.default is inspect.Parameter.empty:
+                        if param in error_ind:
+                            error_ind[param].append(i)
+                        else:
+                            error_ind[param] = [i]
+                    elif not isinstance(self, (PolynomialPhiFunction,
+                                               GaussianPhiFunction)):
+                        if param in warn_ind:
+                            warn_ind[param].append(i)
+                        else:
+                            warn_ind[param] = [i]
```

Comment:
Externalise these parts into the corresponding classes.

Reply:
Removing the ``disable_marginal_guarantee`` argument of the subclasses ``transform`` methods, as you suggested, solved it.